base/runtime_internals.jl

```diff
@@ -173,6 +173,70 @@ false
 """
 ispublic(m::Module, s::Symbol) = ccall(:jl_module_public_p, Cint, (Any, Any), m, s) != 0
 
+_function_macro_error() = (@noinline; error("@__FUNCTION__ can only be used within a function"))
+
+"""
+    @__FUNCTION__
+
+Get the innermost enclosing function object.
+
+!!! note
+    In functions like `f() = [(@__FUNCTION__) for _ in 1:10]`, this would
+    refer to the generator function in the comprehension, NOT the enclosing
+    function `f`. Similarly, in a closure function, this will refer to the
+    closure function object rather than the enclosing function. Note that
+    macros like [`@spawn`](@ref Threads.@spawn), [`@async`](@ref), etc., also
+    create closures.
+
+!!! note
+    This does not work in the context of callable structs as there is no
+    function to refer to, and will result in an `UndefVarError`.
+
+# Examples
+
+`@__FUNCTION__` is useful for closures that need to refer to themselves,
+as otherwise the function object would be captured as a variable and boxed.
+
+```jldoctest
+julia> function make_fib()
+           fib(n) = n <= 1 ? 1 : (@__FUNCTION__)(n - 1) + (@__FUNCTION__)(n - 2)
+           return fib
+       end
+make_fib (generic function with 1 method)
+
+julia> make_fib()(7)
+21
+```
+
+If we had instead used `fib(n - 1) + fib(n - 2)` directly, `fib` would be boxed,
+leading to type instabilities.
+
+Note that `@__FUNCTION__` is also available for anonymous functions:
+
+```jldoctest
+julia> factorial = n -> n <= 1 ? 1 : n * (@__FUNCTION__)(n - 1);
+
+julia> factorial(5)
+120
+```
+
+`@__FUNCTION__` can also be combined with `nameof` to get the symbol for an
+enclosing function:
+
+```jldoctest
+julia> bar() = nameof(@__FUNCTION__);
+
+julia> bar()
+:bar
+```
+"""
+macro __FUNCTION__()
+    quote
+        $(esc(Expr(:isdefined, :var"#self#"))) || $(esc(_function_macro_error))()
+        $(esc(:var"#self#"))
+    end
```

Comment:
nit: much clearer and more efficient to put the esc on the outside instead on every expression:
```suggestion
    esc(quote
        false && (var"#self#" = var"#self#") # declare self is a local variable to avoid polluting or accessing globals
        @isdefined var"#self#" || $Base._function_macro_error()
        var"#self#"
    end)
```